src/cowboy_decompress_h.erl

```diff
@@ -0,0 +1,217 @@
+	-> {cowboy_stream:commands(), #state{}}.
+init(StreamID, Req, Opts) ->
+	RatioLimit = maps:get(decompress_ratio_limit, Opts, 20),
+	Ignore = maps:get(decompress_ignore, Opts, false),
+	State = check_req(Req),
+	Inflate = case State#state.compress of
+		undefined ->
+			undefined;
+		gzip ->
+			Z = zlib:open(),
+			zlib:inflateInit(Z, 31),
+			Z
+	end,
+	{Commands, Next} = cowboy_stream:init(StreamID, Req, Opts),
+	fold(Commands, State#state{next=Next, ratio_limit=RatioLimit, ignore=Ignore,
+		inflate=Inflate}).
+
+-spec data(cowboy_stream:streamid(), cowboy_stream:fin(), cowboy_req:resp_body(), State)
+	-> {cowboy_stream:commands(), State} when State::#state{}.
+data(StreamID, IsFin, Data, State=#state{next=Next0, inflate=undefined}) ->
+	{Commands, Next} = cowboy_stream:data(StreamID, IsFin, Data, Next0),
+	fold(Commands, State#state{next=Next, read_body_is_fin=IsFin});
+data(StreamID, IsFin, Data, State=#state{next=Next0, ignore=true, read_body_buffer=Buffer}) ->
+	{Commands, Next} = cowboy_stream:data(StreamID, IsFin,
+		<< Buffer/binary, Data/binary >>, Next0),
+	fold(Commands, State#state{next=Next, read_body_is_fin=IsFin});
+data(StreamID, IsFin, Data, State0=#state{next=Next0, ratio_limit=RatioLimit,
+		inflate=Z, is_reading=true, read_body_buffer=Buffer0}) ->
+	Buffer = << Buffer0/binary, Data/binary >>,
+	case inflate(Z, RatioLimit, Buffer) of
+		{error, Type} ->
+			Status = case Type of
+				data -> 400;
+				size -> 413
+			end,
+			Commands = [
+				{error_response, Status, #{<<"content-length">> => <<"0">>}, <<>>},
+				stop
+			],
+			fold(Commands, State0#state{inflate=undefined});
+		{ok, Inflated} ->
+			State = case IsFin of
+				nofin ->
+					State0;
+				fin ->
+					zlib:inflateEnd(Z),
+					zlib:close(Z),
+					State0#state{inflate=undefined}
+			end,
+			{Commands, Next} = cowboy_stream:data(StreamID, IsFin, Inflated, Next0),
+			fold(Commands, State#state{next=Next, read_body_buffer= <<>>,
+				read_body_is_fin=IsFin})
+	end;
+data(_, IsFin, Data, State=#state{read_body_buffer=Buffer0}) ->
+	Buffer = << Buffer0/binary, Data/binary >>,
+	{[], State#state{read_body_buffer=Buffer, read_body_is_fin=IsFin}}.
+
+-spec info(cowboy_stream:streamid(), any(), State)
+	-> {cowboy_stream:commands(), State} when State::#state{}.
+info(StreamID, Info, State=#state{next=Next0, inflate=undefined}) ->
+	{Commands, Next} = cowboy_stream:info(StreamID, Info, Next0),
+	fold(Commands, State#state{next=Next});
+info(StreamID, Info={CommandTag, _, _, _, _}, State=#state{next=Next0, read_body_is_fin=IsFin})
+		when CommandTag =:= read_body; CommandTag =:= read_body_timeout ->
+	{Commands0, Next1} = cowboy_stream:info(StreamID, Info, Next0),
+	{Commands, Next} = data(StreamID, IsFin, <<>>, State#state{next=Next1, is_reading=true}),
+	fold(Commands ++ Commands0, Next);
+info(StreamID, Info={set_options, Opts}, State=#state{next=Next0,
+		ignore=Ignore0, ratio_limit=RatioLimit0}) ->
+	Ignore = maps:get(decompress_ignore, Opts, Ignore0),
```

Comment:
Second thoughts on this. We probably shouldn't allow setting `ignore` in the middle of the data, right? If we do the data becomes useless but is still passed forward.

Reply:
OK after some testing we really shouldn't allow setting the `ignore` option after reading has begun. We can still allow `set_options` to set it, but only if there hasn't been data going through yet. I'll fix it.

Reply:
Should be done locally.